build(nix): Package gguf-py

ditsuke opened the pull request:
## What and why
Currently, the Nix build does not ship with the `gguf-py` lib, rendering scripts like `convert-hf-to-gguf` unusable.

This is an attempt to package `gguf-py` with `buildPythonPackage`. Very barebones at the moment and needs some work to exclude the Nix built `gguf` from the devShells, and some thought into other things like packaging the heavier python dependencies (we don't want them to be included by default, but we need a way to opt-in).


CC: @SomeoneSerge

ditsuke committed:
build(nix): Package gguf-py

SomeoneSerge commented on .devops/nix/package.nix:
```diff
@@ -1,35 +1,37 @@
-  vulkan-loader,
-  clblast,
-  useBlas ? builtins.all (x: !x) [
+{ lib
+, config
+, stdenv
+, mkShell
+, cmake
+, ninja
+, pkg-config
```
The previous commit used `github:piegamesde/nixfmt/commits/rfc101-style` for formatting, let's keep this consistent

ditsuke replied:
Thanks, I was wondering how styles are managed. Should we perhaps add a `formatter` to flake.nix?

SomeoneSerge replied:
Probably a good idea. A separate PR maybe?

ditsuke replied:
Will open one soon

ditsuke replied:
#5687

SomeoneSerge commented on .devops/nix/package.nix:
```diff
@@ -62,6 +64,28 @@ let
     strings.optionalString (suffices != [ ])
       ", accelerated with ${strings.concatStringsSep ", " suffices}";
 
+  gguf-py = python3.pkgs.buildPythonPackage rec {
+    pname = "gguf";
+    version = "0.0.0";
+    pyproject = true;
+    nativeBuildInputs = with python3.pkgs; [
+      poetry-core
+    ];
+    # buildInputs = [
+    #   python3.pkgs.poetry-core
+    # ];
+    propagatedBuildInputs = with python3.pkgs; [
+      numpy
+    ];
+    src = lib.cleanSource ../../gguf-py;
+    doCheck = false;
+    meta = with lib; {
+      description = "Python package for writing binary files in the GGU format";
+      license = licenses.mit;
+      maintainers = [ maintainers.philiptaron ];
+    };
+  };
```
1. You don't have to define this in the let-in, instead you can define `gguf-py` in `scope.nix`. Other packages defined in `scope.nix` can just ask for `gguf-py` as an argument
2. Please make a separate file (e.g. `.../gguf-py/default.nix`) and write it as a function accepting `buildPythonPackage` (`numpy`, `poetry-core`, etc) as an argument so you don't hard-code the python version. You can then `gguf-py = callPackage .../gguf-py { }` from `scope.nix`

ditsuke replied:
Cool pattern. I took a dig, can rename/move around the gguf-py package file later.

ditsuke committed:
build(nix): Refactor to new scope for gguf-py

ditsuke committed:
style: format with nixfmt/rfc101-style

ditsuke committed:
build(nix): Exclude gguf-py from devShells

ditsuke commented:
This PR is good for review for the standalone packaging of `gguf-py`. However for the `convert*` scripts that ship with this repo to work, they need a Python environment that has the full superset of dependencies they rely on including `torch`. We'll need to ship with these heavy dependencies, or alternatively we could:
1. Decouple the scripts from default `llama` package, and expose them through a `llama-scripts` package instead.
2. By default have them run against the default python interpreter in the environment and ship the static-environment version as another option.

Open to other ideas.

SomeoneSerge commented on .devops/nix/package-gguf-py.nix:
```diff
@@ -0,0 +1,20 @@
+{
+  lib,
+  llamaVersion,
+  python3,
```
`buildPythonPackage`, `poetry-core`, `numpy` as arguments

ditsuke replied:
The evaluating scope should set them explicitly then?

ditsuke replied:
Not sure if there's a cleaner way: [4957125](https://github.com/ggerganov/llama.cpp/pull/5664/commits/4957125daf29cae8e1ac3d1a81601ca765e080d2)

SomeoneSerge replied:
No, the scope inherits from `python3Packages`, which in turn inherits from `pkgs`

SomeoneSerge replied:
Ah, sorry, mistake on my part. This one only inherits from `pkgs`, we need to handle python separately

SomeoneSerge replied:
If by "good to go" you mean we need to move the python scripts out of the main derivation so that it doesn't depend on the choice of python version 😅

SomeoneSerge commented on .devops/nix/package.nix:
https://github.com/ggerganov/llama.cpp/pull/5664/commits/d63fa07d89047848487e12d4d85fcf9b4e6a3c60 still contains the formatting change, and then there's another commit that flips it back. Would nice to rebase so that the redundant change never happens.

Also I noticed that `nixfmt`'s behaviour concerning the `finalAttrs:` line has changed - so maybe the first commit should be the one for treewide nixfmt, and the later ones can be the functional changes

ditsuke replied:
Good idea, I've rebased to have fmt first

ditsuke committed:
build(nix): Refactor gguf-py derivation to take in exact deps

SomeoneSerge commented on .devops/nix/package-gguf-py.nix:
```diff
@@ -0,0 +1,22 @@
+  nativeBuildInputs = [ poetry-core ];
+  propagatedBuildInputs = [ numpy ];
+  src = lib.cleanSource ../../gguf-py;
+  doCheck = false;
```
Looks like the tests are currently a no-op, but we should enable them and add `pytestCheckHook` in case they are enabled later. We should also add `pythonImportsCheck`

SomeoneSerge commented on .devops/nix/package.nix:
```diff
@@ -58,30 +59,30 @@ let
+    suffices != [ ]
+  ) ", accelerated with ${strings.concatStringsSep ", " suffices}";
+
+  mapToPythonPackages = ps: packages: map (package: ps.${package}) packages;
```
I think this adds complexity and conceals the original issue, which is that we should be using the `substituteInPlace` and `python3.withPackages` for this, and in fact we should separate the python stuff from the C++ derivation

ditsuke replied:
The latest series of commits addresses this separation

ditsuke committed:
build(nix): Enable pytestCheckHook and pythonImportsCheck for gguf-py

ditsuke committed:
build(python): Package python scripts with pyproject.toml

ditsuke committed:
chore: Cleanup

ditsuke committed:
dev(nix): Break up python/C devShells

ditsuke committed:
build(python): Relax pytorch version constraint

Nix has an older version

SomeoneSerge commented on .devops/nix/package.nix:
```diff
@@ -227,21 +210,8 @@ effectiveStdenv.mkDerivation (finalAttrs: {
-      buildInputs = [
-        python3.withPackages
-        (ps: mapToPythonPackages ps llama-python-full-deps)
+        cmake
+        gcc
```
CMake is to be executed on the build machine, it goes in `nativeBuildInputs`.
No need for `gcc` because `stdenv` exposes one.

SomeoneSerge commented on bin/convert_hf_to_gguf.py:
```diff
@@ -22,7 +22,7 @@
 import gguf
 
-from convert import HfVocab
+from bin.convert import HfVocab
```
1. We'll need to check with `@ggerganov` (not pinging yet) about moving stuff around, I don't know what assumptions people might be making about paths...
2. Generally, using top-level names like `bin` or `convert` is rather unhygienic, it's asking for conflicts... With your move a relative import should work, but first...


...what was the motivation for the move?

ditsuke replied:
Ah, having a subdir makes it a tad easier to get the package working with poetry, but certainly not a requirement. I'll revert the locations.

ditsuke replied:
However, we'll still need to convert the filenames from kebab-case to snake_case, otherwise they can't be treated and run as modules.

SomeoneSerge commented on pyproject.toml:
We'd probably do just fine with the setuptools pep-517 back-end... Whatever the backend, the pep517 stuff is probably fit for a separate PR and would be merged faster that way:)

ditsuke replied:
Since `gguf-py` is using pyproject+poetry, perhaps its okay to have that here too?

ditsuke replied:
#5745

ditsuke committed:
chore: Move cmake to nativeBuildInputs for devShell

SomeoneSerge commented on .devops/nix/nixpkgs-instances.nix:
Does nixpkgs' nixfmt-rfc-style in https://github.com/ggerganov/llama.cpp/pull/5687/ allow us to undo the formatting changes?

ditsuke replied:
Apologies, missed this last week. I've rebased and reconciled formatting chances.

ditsuke committed:
fmt: Reconcile formatting with rebase

ditsuke committed:
style: nix fmt

ditsuke committed:
cleanup: Remove unncessary __init__.py

ditsuke commented:
Rebased to after merge of #5745. Commits will need to be cleaned up, but this is open for review otherwise.

SomeoneSerge commented on .devops/nix/python-scripts.nix:
```diff
@@ -0,0 +1,43 @@
+in
+
+buildPythonPackage ({
+  pname = "llama-scripts";
+  src = ../../.;
+  version = "0.0.0";
+  pyproject = true;
+  nativeBuildInputs = [ poetry-core ];
+  projectDir = ../../.;
+  propagatedBuildInputs = llama-python-deps;
```
What's projectDir for?

SomeoneSerge replied:
`src = ../../.` will trigger many unnecessary rebuilds, e.g. when editing nix or yaml files

SomeoneSerge replied:
It's customary to sort the attributes in the phase order and to put `pname version pyproject=true src` at the top.

SomeoneSerge replied:
Nixpkgs might be preparing to switch from build-time to eval-time propagation of python dependencies, and and within that effort it's recommended to use `dependencies = [ ... ]` instead of `propagatedBuildInputs = [ ... ]` (`buildPythonPackage` currently just merges `dependencies` back into `propagatedBuildInputs` so it's an alias, but this might switch to a "smarter" logic in future)

ditsuke replied:
> What's projectDir for?

Totally redundant, removed

> src = ../../. will trigger many unnecessary rebuilds, e.g. when editing nix or yaml files

Good point, I've added filtering for non-source files

> recommended to use dependencies = [ ... ] instead of propagatedBuildInputs = [ ... ]

Nice to know! Done

ditsuke committed:
chore: Suggestions from review

- Filter out non-source files from llama-scripts flake derivation
- Clean up unused closure
- Remove scripts devShell

ditsuke commented on .devops/nix/python-scripts.nix:
```diff
@@ -0,0 +1,43 @@
+  propagatedBuildInputs = llama-python-deps;
+
+  passthru = {
+    shell = mkShell {
+      name = "shell-python-scripts";
+      description = "contains numpy and sentencepiece";
+      buildInputs = llama-python-deps;
+      shellHook = ''
+        addToSearchPath "LD_LIBRARY_PATH" "${lib.getLib stdenv.cc.cc}/lib"
+      '';
+    };
```
I agree that it feels out of place in the derivation. The thought was having this for developing the scripts, but I guess it's not terribly useful? I've removed it for now unless there's a better way to have it.

compilade commented on .devops/nix/package.nix:
```diff
@@ -61,54 +66,13 @@ let
-      ps.torchWithoutCuda
-      ps.transformers
-
-      # server bench
-      ps.matplotlib
-
-      # server tests
-      ps.openai
-      ps.behave
-      ps.prometheus-client
-
-      # for examples/pydantic-models-to-grammar-examples.py
-      ps.docstring-parser
-      ps.pydantic
-
-      # for scripts/compare-llama-bench.py
-      ps.gitpython
-      ps.tabulate
```
I think these new optional dependencies added in #8341 should also be ported over in some way. (to `.devops/nix/python-scripts.nix`?)

(Just so you know, I use `nix develop .#default-extra` a lot in my workflow. What would be the equivalent after this gets merged?)

ditsuke replied:
That's a good question. What are your thoughts about:
1. Removing the passthru.shell/shell-extra construct
2. Adding a shell-only scope that yields a devShell with `inputsFrom` the main package, the llama-scripts package (plus dev dependencies)

cc @SomeoneSerge

compilade replied:
> 1. Removing the passthru.shell/shell-extra construct

I think this construct is not necessary, since `nix develop` can already build a shell from the dependencies of a package. So I agree with removing it.

> 2. Adding a shell-only scope that yields a devShell with `inputsFrom` the main package and the `llama-scripts` package (plus dev dependencies)

This sounds like the right thing to do. All I want is some name to give to `nix develop` to get all the Python dependencies + the C/C++ build dependencies of the default package, like `.#default-extra` currently seems to do.

Ideally, this would make devShells for each of the `llama.cpp` packages (`default-extra`, `vulkan-extra`, `cuda-extra`, etc.).

I think it should be doable to make theses `-extra` shells from `.devops/nix/devshells.nix` as you're suggesting (with `inputsFrom`) without using the `passthru` construct and without necessarily having to make the non-`extra` ones, because the python scripts are now in their own derivation and because `nix develop` seems to do the right thing even when `.devops/nix/devshells.nix` is commented-out, i.e. it makes a shell with the dependencies of the `.#vulkan` package when using `nix develop .#vulkan`.

Unfortunately, `nix develop` can only be given 1 name at a time.
Otherwise I would do `nix develop .#default .#python-scripts`, but this doesn't work, so `-extra` shells seem needed.

ditsuke replied:
@compilade apologies for the delayed turnaround for this. Please take a look at [0188318](https://github.com/ggerganov/llama.cpp/commit/01883187b519240c94c9fb156a870483256acd9a).

compilade replied:
@ditsuke Thanks for having implemented that. I've tried it, and it's nice that it also builds the `gguf` Python package and puts into the environment.

Something I've noticed, though, is that the `$PATH` now contains much more Python dependencies, even though they were also previously accessible to the Python interpreter before that. Not sure if that's a bad thing or not. Hopefully no Python package has a binary called `bash`, because given the search order, it would have priority.

It seems like some of the `gguf-py/scripts` are included in the `gguf` package, which means they are now in my `$PATH` when using `nix develop .#default-extra`. This can be nice to test if the Python build works properly, but it requires re-making the devShell at each change.

This is not a problem, because when directly testing the scripts, they insert the path of the local `gguf` library when `NO_LOCAL_GGUF` is not defined.

So my only remaining concern about this is the Python dependencies in the `$PATH`. But it's not a big problem. Actually, it might be useful because it makes stuff like `transformers-cli` and `huggingface-cli` available too.

I'll try to review this more deeply soon, and hopefully approving it.

ditsuke committed:
revert: Bad changes

ditsuke committed:
dev: Simplify devShells, restore the -extra devShell

ditsuke committed:
build(nix): Add pyyaml for gguf-py

compilade commented on .devops/nix/python-scripts.nix:
```diff
@@ -0,0 +1,66 @@
+
+let
+  llama-python-deps = with python3Packages; [
+    numpy
+    sentencepiece
+    transformers
+    protobuf
+    torchWithoutCuda
+    gguf-py
+    tqdm
```
I see `tiktoken` has been removed from the Python dependencies. It was **not** used by `gguf` or `llama.cpp` directly, but by some models during conversion for their tokenizer (like initially StableLM2 1.6B used `tiktoken`, although they later completely switched to `tokenizer.json` so technically `tiktoken` is no longer necessary for that model in particular).

See <https://github.com/ggerganov/llama.cpp/pull/5052#issuecomment-1902698030> for some background on this.

To be clear, I'm also okay with not including `tiktoken`, because It has been a while since I last saw a model which used this. It was added for convenience, but maybe it's not necessary.

ditsuke replied:
Good spot! I think its a good idea to keep tiktoken in the extra environments if it may be required by a script so I've added it as an additional input to those devShells in [e61d467](https://github.com/ggerganov/llama.cpp/pull/5664/commits/e61d4672dcb5f42fa9140e4b6d1ce6a20d8dac63)

compilade reviewed (approved):
This seems to work for my workflow with `nix develop .#default-extra`. Thanks!

I've tested running `pyright` on the whole repository (which requires access to all python libraries used within), and it works without problem.

I've also tested building a `-DGGML_VULKAN=TRUE` build with `nix develop .#vulkan-extra`, and it works too.

There are more things in the `$PATH` (because python libraries only were added to the python interpreter before, not also the `$PATH`), but I think this might be useful in some situations.

Rebuilding `gguf-py` at each change might be cumbersome, though, but I think this is better than before because I simply never tested building `gguf-py` with poetry before, but with this it's now handled when creating a devShell. (I never needed to actually build the package because `convert_hf_to_gguf.py` uses some other trick anyway to include the local `gguf-py` instead of the package)

I think I like this.

ditsuke committed:
chore: Remove some unused bindings

ditsuke committed:
dev: Add tiktoken to -extra devShells

ditsuke commented:
Rebased to master. @ggerganov this should be ready for merge. cc @SomeoneSerge

ditsuke commented:
Thanks for merging!

SomeoneSerge commented on .devops/nix/scope.nix:
```diff
@@ -1,19 +1,41 @@
 }:
 
+let
+  pythonPackages = python3.pkgs;
```
A note for the future: I'd discourage using `python3.pkgs` in favour of `python3Packages` because the former does not support splicing. We're not using any cross-compilation in this flake, but I think it's important to stay minimal and idiomatic so as not to spread confusion

ditsuke replied:
That's good to know. This should be changed to use python3Packages next time someone visits these files.

ditsuke commented on .devops/nix/scope.nix:
```diff
@@ -1,19 +1,41 @@
 
+let
+  pythonPackages = python3.pkgs;
+  buildPythonPackage = pythonPackages.buildPythonPackage;
+  numpy = pythonPackages.numpy;
+  tqdm = pythonPackages.tqdm;
+  sentencepiece = pythonPackages.sentencepiece;
+  pyyaml = pythonPackages.pyyaml;
+  poetry-core = pythonPackages.poetry-core;
+  pytestCheckHook = pythonPackages.pytestCheckHook;
```
Yes would be sure nice to have a python scope at some point, this is rather ugly.
